<div class="alert alert-info">
<font size="5"><b>Комментарий ревьюера</b></font>

Привет Руслан! Меня зовут Марат, и я буду твоим ревьюером. Спешу сообщить что все ключевые этапы в работе выполнены,  с задачей тебе удалось справиться. По поводу обращения - в IT сфере принято общаться на «ты» :) Но, если привычней на «вы», дай знать. Как ревьюера моя задача помочь тебе в развитии, дав хорошие советы. Я внимательно посмотрю твой код, ознакомлюсь с твоими выводами и оставлю комментарии. Где то могу предложить небольшие исправление в коде, но ненавязчиво. Где потребуются уточнения, я оставлю много наводящих вопросов. Они помогут тебя с поиском верного решения.

Все мои комментарии размечены по цветам, для лучшего восприятия. 
    
<div class="alert alert-success">Зеленым цветом и словом «Успех» отмечены особо удачные и элегантные решения, которыми ты можешь гордиться. </div>
        
<div class="alert alert-warning">Желтым и значком словом «Совет», помечены решения у которых есть альтернативные решения, более оптимальные. Ты можешь найти их сразу и доработать проект, или отложить это на потом, для будущих проектах. Проект будет принят и без их доработки. </div>
        
<div class="alert alert-danger"> Красным цветом и значком словом «Ошибка» помечу твои решения, на которые стоит обратить внимание прежде всего. После их доработки проект будет принят. </div>
        
Залог успеха - работа сообща, взаимное уважение и работа в диалоге. Поэтому, помечай свои ответные комментарии на мои реплики заметным цветом или курсивом, так мне будет легче их отслеживать. Пожалуйста, не изменяй и не удаляй мои комментарии. Все это поможет выполнить повторную проверку быстрей.

</div>

# Рекомендация тарифов

В вашем распоряжении данные о поведении клиентов, которые уже перешли на эти тарифы (из проекта курса «Статистический анализ данных»). Нужно построить модель для задачи классификации, которая выберет подходящий тариф. Предобработка данных не понадобится — вы её уже сделали.

Постройте модель с максимально большим значением *accuracy*. Чтобы сдать проект успешно, нужно довести долю правильных ответов по крайней мере до 0.75. Проверьте *accuracy* на тестовой выборке самостоятельно.

# Описание данных 
Каждый объект в наборе данных — это информация о поведении одного пользователя за месяц. Известно:
* сalls — количество звонков,
* minutes — суммарная длительность звонков в минутах,
* messages — количество sms-сообщений,
* mb_used — израсходованный интернет-трафик в Мб,
* is_ultra — каким тарифом пользовался в течение месяца («Ультра» — 1, «Смарт» — 0).

# План работы:
* 1 Откройте и изучите файл
* 2 Разбейте данные на выборки
* 3 Исследуйте модели
* 4 Проверьте модель на тестовой выборке
* 5 (бонус)Проверьте модели на адекватность


<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех:


Вступление в работу очень важно, так человек, который смотрит твой проект (и на работе в том числе) будет сразу введен в курс дела.





</div>

## Откройте и изучите файл

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import mean_squared_error
from sklearn.dummy import DummyClassifier

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех:

Собираем все импорты в верхней части, чтобы легче было ориентироваться и добавлять новые по необходимости. 


<div class="alert alert-warning">

Совет:     
    
    


- Руслан, у тебя тут есть лишние импорты, их стоит убрать, чтобы поберечь ресурсы  

- кстати есть рекомендации PEP-8 при написании кода, в том числе и для импортов. Если интересно можешь почитать [тут](https://pythonworld.ru/osnovy/pep-8-rukovodstvo-po-napisaniyu-koda-na-python.html), в будущем пригодится )  



<div>

<div class="alert alert-info" style="border-radius: 10px; box-shadow: 2px 2px 2px; border: 1px solid; padding: 10px "> 
<b>Комментарий студента
    
Лишнее удалил 
    </b> 

<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюераV2</b></font>



Совет 🤔:

    
Привет Руслан!    
    
А вот ещё лишнее
    
    from sklearn.metrics import mean_squared_error

In [2]:
data = pd.read_csv('/datasets/users_behavior.csv')
data

calls  minutes  messages   mb_used  is_ultra
0      40.0   311.90      83.0  19915.42         0
1      85.0   516.75      56.0  22696.96         0
2      77.0   467.66      86.0  21060.45         0
3     106.0   745.53      81.0   8437.39         1
4      66.0   418.74       1.0  14502.75         0
...     ...      ...       ...       ...       ...
3209  122.0   910.98      20.0  35124.90         1
3210   25.0   190.36       0.0   3275.61         0
3211   97.0   634.44      70.0  13974.06         0
3212   64.0   462.32      90.0  31239.78         0
3213   80.0   566.09       6.0  29480.52         1

[3214 rows x 5 columns]

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех:

👍 Данные изучены.

<div class="alert alert-warning">



Совет: 


- .info (дает тебе информацию о пропусках и типах данных), а .describe (медианы, квантили, max, min итд) это такой джентльменский набор при первичной разведке данных (EDA).  



- Можно еще отдельно проверить датасет на сбалансированность классов в таргете. Это помогло бы решить бонусное задание сразу (Только не надо балансировать данные, это тема следующего проекта).




- Можно посмотреть корреляцию признаков. Знаешь что такое мультиколлинеарность, какие два типа проблем возникает, для каких моделей и какие варианты решения?






- сейчас это не нужно (пока лучше вручную набивать все эти .info, count_values, .describe, .diplicate...) но в будущем можешь обратить внимание на библиотечки, в которых уже собран необходимый минимум при EDA (и не только минимум), например [pandas_profiling](https://habr.com/ru/companies/ruvds/articles/451478/) - будет информативно и красиво

## Разбейте данные на выборки

**Отделим целевой признак от остальных признаков**

In [4]:
features= data.drop(['is_ultra'], axis=1)
target = data['is_ultra']
display(features.head(5)) # проверим разделение
display(target.head(5))

calls  minutes  messages   mb_used
0   40.0   311.90      83.0  19915.42
1   85.0   516.75      56.0  22696.96
2   77.0   467.66      86.0  21060.45
3  106.0   745.53      81.0   8437.39
4   66.0   418.74       1.0  14502.75

0    0
1    0
2    0
3    1
4    0
Name: is_ultra, dtype: int64

<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюера</b></font>



Совет 🤔:

print не всегда нужен, вывод часто произойдет и так, а когда нужен, вместо print лучше использовать display, будет красивей





</div>

<div class="alert alert-info" style="border-radius: 10px; box-shadow: 2px 2px 2px; border: 1px solid; padding: 10px "> 
<b>Комментарий студента
    
Применил (display())
    </b> 

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюераV2</b></font>



Успех 👍:



Симпатичнее




</div>


**Разделим исходные данные на обучающую, валидационную и тестовую выборки.**

Разделим данные в соотношении 60/20/20 на обучающая/валидационну/тестовую выборку.

**Обучающая выборка**

In [5]:
features_train, features_1, target_train, target_1 = train_test_split(\
                                                                      features,\
                                                                      target, \
                                                                      test_size=0.4,\
                                                                      random_state=12345) 

In [6]:
#проверяю разбивку
print(features_train.shape)
print(target_train.shape) 
print(f'Обучающая выборка: {features_train.shape[0] / data.shape[0] * 100}% от data')

(1928, 4)
(1928,)
Обучающая выборка: 59.98755444928439% от data


**Валидационная и тестовая выборка**

In [7]:
features_valid, features_test, target_valid, target_test = train_test_split(\
                                                                            features_1,\
                                                                            target_1,\
                                                                            test_size=0.5,\
                                                                            random_state=12345) 

In [8]:
#проверяю разбивку
print(f'Валидационная выборка features:{features_valid.shape}')
print(f'Валидационная выборка target:{target_valid.shape}')
print(f'Тестовая выборка features:{features_test.shape}')
print(f'Тестовая выборка target:{target_test.shape}')
print(f'Валидационная выборка: {features_valid.shape[0] / data.shape[0] * 100}% от data')
print(f'Тестовая выборка: {features_test.shape[0] / data.shape[0] * 100}% от data')

Валидационная выборка features:(643, 4)
Валидационная выборка target:(643,)
Тестовая выборка features:(643, 4)
Тестовая выборка target:(643,)
Валидационная выборка: 20.00622277535781% от data
Тестовая выборка: 20.00622277535781% от data


<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех:

Все правильно!

Здорово что проконтролировано разбиение (можно с помощью .shape,  можно .len)



<div class="alert alert-warning">



Совет: 


- Обрати внимание на аргумент stratify, он позволит сохранить изначальное распределение таргетов во всех новых датасетах.  Существующий дисбаланс никуда не денется, но в каждом датасете он будет одинаковым. [Почитать](https://pythonru.com/baza-znanij/sklearn-train-test-split) можно тут






## Исследуйте модели

### **DecisionTreeClassifie (Деревья решений)**

In [9]:
for depth_1 in range(1,11):
    model_1 = DecisionTreeClassifier(random_state=12345, max_depth=depth_1) 
    model_1.fit(features_train, target_train)
    predictions_valid_1 = model_1.predict(features_valid)
    print('max_depth =', depth_1, ':', end='')
    print(accuracy_score(target_valid, predictions_valid_1))

max_depth = 1 :0.7542768273716952
max_depth = 2 :0.7822706065318819
max_depth = 3 :0.7853810264385692
max_depth = 4 :0.7791601866251944
max_depth = 5 :0.7791601866251944
max_depth = 6 :0.7838258164852255
max_depth = 7 :0.7822706065318819
max_depth = 8 :0.7791601866251944
max_depth = 9 :0.7822706065318819
max_depth = 10 :0.7744945567651633


**Вывод:** Accuracy лучшей модели равно 0.7853810264385692, с гиперпараметром max_depth равным - 3.

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех:
 
Работа выполнена в соответствии с критериями: 



 - модель обучена на обучающем наборе
 - получена оценка качества на валидационном наборе
 - перебор гиперпараметров осуществляется в цикле





<div class="alert alert-warning">


Совет: 




- Можно добавить график, хорошо оформленный график украсит проект. Тут на графике можно посмотреть как при изменении max_depth меняется точность на train и validation датасете (для этого собираем наши метрики на валидации и train в списки, а затем по ним обычный plt.plot()) Такой способ используют чтобы лучше отследить переобучение, но для этого стоит взять пернебор по max_depth поглубже


- tree_plot можно построить, с max_depth равной 3 или 4 (глубже уже будет громоздко) - глянуть как DT сплиты делает, будет понимание как модель принимает решения: какие признаки  использует, на каких значениях делает разбиение вправо - влево 




</div>



### **RandomForestClassifier(Случайный лес)**

In [17]:
%%time
best_model_2 = None
best_est_2 = 0
best_depth_2 = 0
best_result_2 = 0
for est in range(1, 11):
    for depth in range(1, 11):
        model_2 = RandomForestClassifier(random_state=12345, n_estimators=est, max_depth = depth)
        model_2.fit(features_train, target_train)
        predictions_valid_2 = model_2.predict(features_valid)
        result_2 = accuracy_score(target_valid, predictions_valid_2)
        if best_result_2 < result_2:
            best_model_2 = model_2
            best_est_2 = est
            best_depth_2 = depth
            best_result_2 = result_2
            
print('Лучший гиперпараметр n_estimators =', best_est_2)
print('Лучший гиперпараметр max_depth =', best_depth_2)
print('Accuracy наилучшей модели',best_result_2)

Лучший гиперпараметр n_estimators = 8
Лучший гиперпараметр max_depth = 8
Accuracy наилучшей модели 0.80248833592535
CPU times: user 1.77 s, sys: 20.9 ms, total: 1.79 s
Wall time: 1.79 s




<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех:
   


 
По большему числу гиперпараметров перебор -   тем лучше метрика. Случайный лес мощная модель и по ней действительно стоит поперебирать несколько гиперпораметров. Здорово   




<div class="alert alert-warning">


Совет: 




- Когда что то долго крутиться, можно использовать  %%time - ставишь на самый вверх ячейки с кодом, время выполнения которого хочешь замерить, может не знаешь.  Быстрее не станет, но все будут видеть стоит ли ждать не отходя от ПК или можно сходить чаек поставить )) 
    
    Или tqdm, это ещё лучше, потому что он показывает интерактивно, на каком этапе расчетов мы находимся

    
    from tqdm import tqdm


    for n_estimators in tqdm(range(3,58)):

        ..........



(дальше мы будем уходить от неэффективных в питоне циклов к apply.  Там тоже есть аналог прогресс-баров:  .progress_apply)
    
И будет красиво )   
 





</div>


<div class="alert alert-info" style="border-radius: 10px; box-shadow: 2px 2px 2px; border: 1px solid; padding: 10px "> 
<b>Комментарий студента
    
Применил (%%time)
    </b> 

### Logistic Regression (Логистическая регрессия)

In [12]:
model_3 = LogisticRegression(random_state=12345)
model_3.fit(features_train, target_train)
predictions_valid_3 = model_3.predict(features_valid)
accuracy_3 = accuracy_score( target_valid, predictions_valid_3)
print('Accuracy логистической регрессии', accuracy_3, end ='')

Accuracy логистической регрессии 0.7107309486780715

**3.Вывод:**
* Accuracy лучший результат из выбронных моделий у RandomForestClassifier(Случайный лес), равен 0.80248833592535 с гиперпараметрами n_estimators = 8 и max_depth = 8
* Accuracy Logistic Regression (логистической регрессии) 0.7107309486780715, эта модель не подходит для задания она ниже 0.75
* Accuracy DecisionTreeClassifie (Деревья решений) равен 0.7853810264385692, с гиперпараметром max_depth равным - 3.

## Проверьте модель на тестовой выборке

Наилучший результат показала RandomForestClassifier(Случайный лес), проверим ее на тестовой выборке.


<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех:


Все верно, RF лучший. 


<div class="alert alert-warning">


Совет: 

Если есть желание можешь ответить на вопросики ))    



- Как назвать ситуацию, когда на валидации при увеличении глубины дерева (модель DT), метрика качества стала снижаться? Как по твоему в это же время вела себя эта же метрика на train?


- Насколько знаю вы еще под капот моделям не заглядывали, но может знаешь почему обычно RF показывает более высокие результаты из выбранных?


- Мы решаем задачу классификации (а еще есть задача регрессии), а в названии модели с помощью которой мы решаем задачу "классификации" (Логистическая регрессия) есть слово "регрессия".  Нет ли тут парадокса? )
    

- Почему логистическая регрессия может показывать на много более худшие результаты?  


</div>





In [13]:
model_4 = RandomForestClassifier(random_state=12345, n_estimators = 8, max_depth = 8)
model_4.fit(features_train, target_train)
predictions_test_4 = model_4.predict(features_test)
result_test_4 = accuracy_score(predictions_test_4, target_test)
print('Accuracy', result_test_4)

Accuracy 0.7962674961119751


**4.Вывод:**
* Accuracy RandomForestClassifier(Случайный лес) на тестовой выборке 0.7962674961119751.


<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех:


- Все верно, логика моделирования не нарушена, тут тестируем только лучшую модель (или парочку, если метрики были близки) отобранную на валидации. 

- Если студент получил на тесте accuraсy  выше 0,78, это считается отличным результатом. Ты подбираешь лучшую комбинацию не по одному гиперпараметру и вот он результат!


<div class="alert alert-warning">

Вопросики и совет:




- Как попробовать улучшить результат, не мучаясь с кодом?! Смотри - мы сформировали train выборку на которой обучаем модель, а валидационную используем для поиска лучших значений гиперпараметров. Оке, нашли их. Так почему бы теперь модель с выбранными гиперпараметрами не обучить на тренировочной + валидационной выборке (только не нужно делать новые сплиты, используй pd.concat(), к примеру)?! Чем больше данных, тем лучше модель может обучиться. И эту дообученную модель проверим на тесте. Гаранитий что станет лучше нет, но ML это постоянные эксперименты.  
    
    
- А можешь предложить как нам использовать полученный результат в бизнесе? Как нам на этих прогнозах сделать деньги? 




- Руслан, в качестве бонуса можешь вывести еще несколько метрик своей модели - precision, recall, F1 и confussion matrix) Все есть в sklearn)


    
- Если ты не будешь дообучать модель, поможешь заметить что лучшую модель ты уже записал тут  -  best_model_2, поэтому это уже не нужно  
    
    
    model_4 = RandomForestClassifier(random_state=12345, n_estimators = 8, max_depth = 8)
    model_4.fit(features_train, target_train)


## (бонус) Проверьте модели на адекватность

Проверяем с помощью модели DummyClassifier

In [14]:
model_5 = DummyClassifier(strategy='most_frequent', random_state=12345)
model_5.fit(features_train, target_train)
result_5 = model_5.score(features_valid, target_valid)
print('Accuracy DummyClassifier:', result_5)

Accuracy DummyClassifier: 0.7060653188180405


**4.Вывод:**
* Accuracy модели RandomForestClassifier(Случайный лес) выше Accuracy DummyClassifier, значит модель адекватна. 

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех:

Все верно. 

👍 что используешь специальную обертку для проверки на адекватность


## Общий вывод

* **Разбили данные на выборки:**
    * на обучающую, валидационную и тестовую в соотношении 60/20/20
* **Проведено исследование моделей:**
    * DecisionTreeClassifie (Деревья решений)
    * RandomForestClassifier(Случайный лес)
    * Logistic Regression (Логистическая регрессия)

* Accuracy лучший результат из выбронных моделий у RandomForestClassifier(Случайный лес), равен 0.80248833592535 с гиперпараметрами n_estimators = 8 и max_depth = 8
* Accuracy Logistic Regression (логистической регрессии) 0.7107309486780715, эта модель не подходит для задания она ниже 0.75
* Accuracy DecisionTreeClassifie (Деревья решений) равен 0.7853810264385692, с гиперпараметром max_depth равным - 3.
   
* **Проверена модель на тестовой выборке:**
    * Наилучший результат показала RandomForestClassifier(Случайный лес), проверим ее на тестовой выборке
    * Accuracy RandomForestClassifier(Случайный лес) на тестовой выборке 0.7962674961119751
* **Проверена модель на адекватность:**
    * Проверяем на адекватность с помощью модели DummyClassifier
    * Accuracy модели RandomForestClassifier(Случайный лес) выше Accuracy DummyClassifier, значит модель адекватна.  
    
* Самая лучшая модель для данной задачи RandomForestClassifier(Случайный лес)


<div class="alert alert-info">
<font size="5"><b>Комментарий ревьюера</b></font>




Руслан, у тебя старательно выполненная работа, все четко, осмысленно. Выводы присутствуют, с комментированием кода тоже никаких проблем нет, твоих объяснений достаточно для понимания коллегами хода твоих мыслей. 

    
    


Я оставил небольшие советы и вопросики (если есть время и желание можешь воспользоваться/ответить): 




- обрати внимание на проверку сбалансированность классов в таргете в разделе EDA
- использовать stratify
- добавить график метрики    
- как можно улучшить результаты 



Посылаю на вторую итерацию, чтобы ответить на вопросы если они появятся



<div class="alert alert-info">
<font size="5"><b>Комментарий ревьюераV2</b></font>

Спасибо за работу! Небольшие изменения внесены    

    

Красного нет, вопросов нет, значит все, пора принимать) Надеюсь мои советы и вопросики были полезны и в копилочку знаний упало что то новое, а проект стал лучше, и симпатичней.

  
Отличная работа Руслан. Желаю успехов в дальнейшей учебе!


## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x] Jupyter Notebook открыт
- [x] Весь код исполняется без ошибок
- [x] Ячейки с кодом расположены в порядке исполнения
- [x] Выполнено задание 1: данные загружены и изучены
- [x] Выполнено задание 2: данные разбиты на три выборки
- [x] Выполнено задание 3: проведено исследование моделей
    - [x] Рассмотрено больше одной модели
    - [x] Рассмотрено хотя бы 3 значения гипепараметров для какой-нибудь модели
    - [x] Написаны выводы по результатам исследования
- [x] Выполнено задание 3: Проведено тестирование
- [x] Удалось достичь accuracy не меньше 0.75
